### PR TITLE
WIP: Implement stable symbol-name generation algorithm.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR changes the way symbol names are generated by the compiler. The new algorithm reflects the current state of the discussion over at https://github.com/rust-lang/rfcs/issues/689.

Once it is done, it will also fix issue https://github.com/rust-lang/rust/issues/30330. I want to add a test case for that before closing it though.

I also want to do some performance tests. The new algorithm does a little more work than the previous one due to various reasons, and it might make sense to adapt it in a way that allows it to be implemented more efficiently.

@nikomatsakis: It would be nice if there was a way of finding out if a `DefPath` refers to something in the current crate or in an external one. The information is already there, it's just not accessible at the moment. I'll probably propose some minor changes there, together with some facilities to allow for accessing `DefPaths` without allocating a `Vec` for them.

**TODO**
- ~~Actually "crate qualify" symbols, as promised in the docs.~~
- ~~Add a test case showing that symbol names are deterministic~~.
- Maybe add a test case showing that symbol names are stable against small code changes.

~~One thing that might be interesting to the @rust-lang/compiler team:
I've used SipHash exclusively now for generating symbol hashes. Previously it was only used for monomorphizations and the rest of the code used a truncated version on SHA256. Is there any benefit to sticking to SHA? I don't really see one since we only used 64 bits of the digest anyway, but maybe I'm missing something?~~ ==> Just switched things back to SHA-2 for now.

### CODE REVIEW

**Comment by bluss** [comment]
SipHash it is an algorithm designed by cryptographers, but it's not in the same category of hash functions as SHA-2 which is a [_cryptographic hash function_.](https://en.wikipedia.org/wiki/Cryptographic_hash_function)

For example, the [SipHash paper](https://131002.net/siphash/) says

> In general, building a MAC from a general-
purpose cryptographic hash function appears to be a highly suboptimal ap-
proach: general-purpose cryptographic hash functions perform many extra com-
putations for the goal of collision resistance on public inputs, while MACs have
secret keys and do not need collision resistance.

and

> We comment that SipHash is not meant to be, and (obviously) is not, collision-
resistant.

This use case does not use a secret key with SipHash (it always uses the default key), and it wants to have collision resistance on public inputs. Yes, in practice, we can find collisions for either of them if the hash is only 64 bits long, but only SHA-2 is designed for this case.

**Comment by michaelwoerister** [comment]
@bluss Those are good references. It really sounds like SipHash is not a good candidate for this use case. I'll switch things to SHA-2.

**Comment by bluss** [comment]
What's the impact of a collision?

**Comment by michaelwoerister** [comment]
> What's the impact of a collision?

Symbol conflicts: Two functions with different signatures could end up with the same name. If it's in the same crate, compilation will probably just fail. If it's in different crates, either linking fails due to a conflict or the wrong symbol might be linked, resulting most probably in untraceable, subtle memory corruption.

**Comment by nikomatsakis** [comment]
Nice.

**Comment by nikomatsakis** [comment]
cc @alexcrichton @brson

**Comment by alexcrichton** [comment]
This looks fantastic to me, nice work @michaelwoerister!

I wouldn't have too many opinions about SHA2 vs SipHash, I suspect either choice will be suitable for our purposes. I'd personally love to jettison the SHA2 implementation in the compiler, and I *highly* doubt we can find a SipHash practical collision with the same path, different metadata, and yet something like different type substitutions. We should also be able to change this if we ever need to really as we can only guarantee no recompiles within one version of Rust anyway.

**Comment by michaelwoerister** [comment]
Thanks for the comment @alexcrichton. It's true, we have been using SipHash for function monomorphization for quite a while now and I've never seen a problem with that.

My assumption about the requirements for this hash function (and later also one used for fingerprinting incremental compilation artifacts) is that it should provide good distribution so that collision probability is as low as it can theoretically be for a given number of hash value bits, BUT that we don't care about  "security", i.e. whether it's easy to find hash collisions on purpose. Or is there some kind of "attack scenario" that I'm overlooking?

For maximum performance (and if the above assumptions hold), we could also use something like [CRC64](https://matt.sh/redis-crcspeed).
Anyway, I think for this PR, I'll just leave SHA-2 in.

**Comment by bluss** [comment]
Here's the issue about collisions in type_id, but that's only tangential to this: #10389. As long as we go into this with the knowledge that SipHash is just a regular hash function, I'm fine.

**Comment by michaelwoerister** [comment]
@bluss Thanks for the link!

**Comment by michaelwoerister** [comment]
As an aside, I think "-C metadata" is a bit of a misnomer. Something like "-C salt" or "-C disambiguator" would be clearer (although the later one definitely is too long).

**Comment by michaelwoerister** [comment]
Another side note: Now that we get predictable symbol names for monomorphizations, we might want to start sharing them cross-crate.

**Comment by alexcrichton** [comment]
Ah as another data point in Cargo we extensively use SipHash for fingerprinting crates and packages. I haven't at least seen anyone run into any problems with that.

I also agree that at this point `-C metadata` is basically just wrong, but I don't think we can rename due to the extensive usage in Cargo right now :(

**Comment by arielb1** [comment]
What's the problem with the compiler SHA2 implementation?

**Comment by michaelwoerister** [comment]
@arielb1 I don't think anything is wrong with the SHA-2 implementation per se. It's just incomplete (only supports SHA256) and it's one more piece of code that needs to be maintained. I'd guess that second point is the main motivation for wanting to get rid of it.

**Comment by michaelwoerister** [comment]
So, here's something interesting: Some of the test cases (e.g. `run-pass/typeid-intrinsic.rs`) link to an external crate that has exactly the same name, and thus the crate graph ends up with two identical crates in it. So far, this was no problem because the two crates still have a different SVH but under the new scheme, two crates are considered the same if they have the same `name/metadata` pair.

The tests now fail with 'duplicate symbol' linker errors, but I guess the sensible thing to do here, is to emit a compilation error when the crate loader discovers two crates with the same name/metadata but different SVHs.

**Comment by michaelwoerister** [comment]
Up until that last commit this passes `make check` for me locally, *with* the last commit it looks good too, except that the `run-pass/backtrace.rs` test case fails (it seems to panic while trying to write the backtrace). Does anybody have an idea what that could be about? In general, creating backtraces seems to work just fine...

**Comment by michaelwoerister** [comment]
That's the error in question: https://travis-ci.org/rust-lang/rust/builds/109449298#L6591

**Comment by alexcrichton** [comment]
Oh you should just be able to update the test to check for `backtrace::foo` instead of only `foo`, I believe that previously the crate name was never on symbols, just in the debug description of symbols.

**Comment by michaelwoerister** [comment]
@alexcrichton I'm afraid it's something more complicated than that. If you look at the actual backtrace that the subprocess provides (see below), it looks like there is a panic *while* creating the backtrace (in `std::sys::backtrace::tracing::imp::write` in gcc_s.rs). Something causes an error there.

```
error: test run failed!
status: exit code: 101
command: x86_64-unknown-linux-gnu/test/run-pass/backtrace.stage2-x86_64-unknown-linux-gnu 
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread '<main>' panicked at 'bad output: thread '<main>' panicked at 'explicit panic', /home/travis/build/rust-lang/rust/src/test/run-pass/backtrace.rs:23
stack backtrace:
   1:     0x2b4313b5fec0 - std::sys::backtrace::tracing::imp::write::h5cb8831505517f1a
   2:     0x2b4313b68e3b - std::panicking::default_handler::_$u7b$$u7b$closure$u7d$$u7d$::h7d4838ac3001375c
   3:     0x2b4313b68a08 - std::panicking::default_handler::heb8f66506b25fc53
   4:     0x2b4313b31ac6 - std::sys_common::unwind::begin_unwind_inner::h896acc46de4d4a1c
   5:     0x2b4313671e44 - std::sys_common::unwind::begin_unwind::h72328ba1025302f0
                        at src/libstd/sys/common/unwind/mod.rs:236
   6:     0x2b4313671dd3 - backtrace::foo::h2bfe25e162df4c7c
                        at /home/travis/build/rust-lang/rust/<std macros>:3
   7:     0x2b43136732a5 - backtrace::main::h1e5dc0fe5e6f026b
                        at src/test/run-pass/backtrace.rs:98
   8:     0x2b4313b68474 - std::sys_common::unwind::try::try_fn::hbb3b06ed012dffa6
   9:     0x2b4313b5d82b - __rust_try
  10:     0x2b4313b680b1 - std::rt::lang_start::ha4a60ae3cf761e9f
  11:     0x2b431424eec4 - __libc_start_main
  12:     0x2b4313671c28 - <unknown>
  13:                0x0 - <unknown>
', /home/travis/build/rust-lang/rust/src/test/run-pass/backtrace.rs:54
note: Run with `RUST_BACKTRACE=1` for a backtrace.

------------------------------------------
```

**Comment by alexcrichton** [comment]
Hm, that looks normal to me though? It looks like the assertion message originates [from here](https://github.com/rust-lang/rust/blob/94ecd483e5ed692a6424b882b706e6615efa8336/src/test/run-pass/backtrace.rs#L54) which was testing a panic [originating here](https://github.com/rust-lang/rust/blob/94ecd483e5ed692a6424b882b706e6615efa8336/src/test/run-pass/backtrace.rs#L23). The bad output happened because it expected to find `- foo` where it instead printed `- backtrace::foo`.

In other words we should probably really look into skipping the top few frames when printing a backtrace, but other than that it should be good!

**Comment by michaelwoerister** [comment]
Yes, you're right of course `:)`
The test passes just fine now.

**Comment by michaelwoerister** [comment]
So, commit https://github.com/rust-lang/rust/commit/545cc8f12111a30771d8ecf26dd25202f06bb232 adds an error to the compiler that is meant to make it easier to understand the new kinds of linker errors that are possible now with the new symbol naming scheme.
However, the condition on which the error is emitted is more conservative than it strictly needs to be:
When the graph of crates being compiled contains two different crates with the same 'crate-name' and the same set of `-Cmetadata` provided salts, the compiler errors out, since then symbol conflicts *can* occur. In many cases though, there wouldn't be any conflicts because the crates don't define the same items.

At the moment we allow these crate graphs, so adding this error will constitute a breaking change. I think, it's a good idea to be conservative here, since it semantically *is* a conflict trying to compiler two different crates with the same "identifier", but I wanted to make sure that this is indeed what we want to do here.

**Comment by alexcrichton** [comment]
Good idea @michaelwoerister! I wonder though, as evidenced by the large number of test renames, maybe we want to somehow consider `crate-type` as well? I think that this may actually affect Cargo as well because we'll commonly have a project with both a library and a binary of the same crate name and also both of which have the same `-C metadata`. In retrospect perhaps not the best forward thinking there, but could mean that this could break in practice?

**Comment by michaelwoerister** [comment]
Hm, it seems a bit ad hoc to make the `crate-type` part of the crate identifier. In the Cargo case you talked of, will the library and binary versions of a crate ever show up in the same crate graph? If so, couldn't Cargo take care of incorporating the `crate-type` by specifying an additional corresponding `-C metadata` option? Do we give any guarantees as to with which `-Cmetadata` values a crate will end up with?

The reason why I'm a bit reluctant to hard-coding the `crate-type` into symbol hashes, is that right now we have a very simple and straightforward model: a crate is identified by its name and, if further disambiguation is needed, the salt provided via `-Cmetadata`.  The salting mechanism is so general that it can be used in place of any other mechanism, and so it feels to me like we are unnecessarily muddling things if we start to incorporate other things into the salt as well.

**Comment by alexcrichton** [comment]
Yeah it's true that it feels ad-hoc, but this comes up pretty frequently I believe because binary crates in Cargo often link to their library counterparts. Cargo could definitely pass `-C metadata`, but currently all existing `cargo` binaries don't. It would basically mean that an older Cargo could not work with a newer compiler (which Cargo currently relies on for snapshots).

We could always land a Cargo change *now* and then hold off on this for a bit? I think this may want to land sooner than that though?

I do agree that the model in this PR *feels* right, although in terms of usability it may lack a bit?

**Comment by alexcrichton** [comment]
Another possibility would be to just say "the salt has `bin` appended to it if a binary is being created" or something like that, but that's also not... great

**Comment by michaelwoerister** [comment]
We do allow one compiler run to produce multiple crate-types, right? I.e. `rustc --crate-type=lib,bin` will produce an executable *and* a library. However, now we have a problem: the executable and the binary should have different salts and thus all their symbol names should be different. It's doable, but implementing this without unnecessarily redoing a lot of work would be a bit of effort (we'd need to keep a map of symbol names and do substitutions in the optimized IR or do that directly in the binaries and in crate metadata).

Another way to handle things would probably be to keep using the old symbol naming scheme and activating the new one only when `-Zincr-comp` is specified. That way we'd have some time to make the change to Cargo without being blocked on it.

**Comment by alexcrichton** [comment]
Yeah you can emit both a bin and a lib at the same time, but that may be sufficiently rare and "weird" that we shouldn't worry about it too much. In either case I don't think we'd want to retranslate everything with just different symbol names.

I'm a little worried about breaking older Cargo's compatibility with newer rustc, though, mostly in terms of portability. Platforms which aren't tier 1 for us and don't have Cargo typically rarely update Cargo as it's somewhat difficult for them to compile, so this'd force them all to update to a newer Cargo. Not necessarily the end of the world, and may not be the much effort in the grand scheme of things.

**Comment by nikomatsakis** [comment]
I personally find these two models equivalent in terms of their
"simplicity":

1. Hash is crate name and salt
2. Hash is crate name, type, and salt

In my mind, the salt logically means "authority and version number".
Basically it is the global naming scheme.

It seems to me that since cargo allows a binary to link to itself as a
library, that implies that option 2 is the right way to think of things.
That is, the salt here would be " crates.io/version" and that remains
constant between both bin/lib. (The actual salt may be different but that
doesn't really matter, this would be sufficient.)

**Comment by michaelwoerister** [comment]
> In my mind, the salt logically means "authority and version number".

I see the salt more as a general purpose disambiguation mechanism (which cargo then uses to provide "authority and version number")

> Another possibility would be to just say "the salt has bin appended to it if a binary is being created" or something like that, but that's also not... great

I'll give this a try now, so that the PR can move forward. At some point, I'd still like to come back to this though and see if we can't do something with more consistent semantics -- maybe at the cost of forcing some non-tier 1 platforms to re-compile their Cargo distributions.

**Comment by alexcrichton** [comment]
I've opened up https://github.com/rust-lang/cargo/pull/2420 to ensure that Cargo generates distinct metadata for libs/bins to hopefully protect against whatever we decide to to moving forward.

**Comment by michaelwoerister** [comment]
To give a little update on this: The core changes to symbol naming seem to be working for some time now, but there are many small issues with the test framework and how it sets up LD_LIBRARY_PATH and related things. It's very time-consuming to work through these issues since some of them only show up in stage2. But (slow) progress is being made on this PR.

**Comment by briansmith** [comment]
The use of the term "salt" is confusing for crypto people reading these patches since this isn't a salt in the sense crypto people normally use. I recommend using some other term.

it seems you are using SHA-256 for some security-sensitive reason. If so, then you should document why/how the inputs to SHA-256 are securely delimited. But, instead, I recommend that you switch to labeled length-prefixed values.

Bad (the way the patch is written):
```
for s in &metadata {
        hasher.input_str(&s[..]);
}
```

Good:
```
hasher.input_str("metadata");
for s in &metadata {
        // Made up function that can hash a 64-bit value.
        hasher.input(s.len() as u64);

        hasher.input_str(&s[..]);
}
```

To see why this is bad, consider `s = { "ab, "c" }` vs `s = { "a", "bc" }`. These would hash to the same value without the length prefixes, but they mean different things. Similarly, instead of having an optional field for "-exe", it is better to have a non-optional, labeled, length-prefixed field.

Also, the documentation for `session.opts.cg.metadata` and other things should be documented so that people maintaining it know that it is a security-relevant function. In particular, all the stuff about labels, length prefixes, and avoiding optional fields would apply to that above.

Finally, the documentation for this code should make it clear exactly why it is secure and what one would need to do to extend it securely (i.e. using labeled length-prefixed subfields, avoiding optional fields, and never reusing a label for a different purpose).

**Comment by briansmith** [comment]
> Ah as another data point in Cargo we extensively use SipHash for fingerprinting crates and packages.

IMO, you should either switch to CRC-32 (if it doesn't need to be secure) or a secure hash function, if it does need to be secure. Misusing SipHash for something it wasn't intended for is worse than either of those options, IMO.

**Comment by michaelwoerister** [comment]
@briansmith Thanks a lot for your comments!
I have some questions:

> The use of the term "salt" is confusing for crypto people reading these patches since this isn't a salt in the sense crypto people normally use. I recommend using some other term.

Can you elaborate on that? The value stored as "crate salt" is used as additional input to the hash function we use when generating symbol hashes. I thought that would kind of match up with how the term is used normally.

> it seems you are using SHA-256 for some security-sensitive reason.

No, we actually don't. At least as far as I know we just want something that provides good random distribution.

> But, instead, I recommend that you switch to labeled length-prefixed values.

That's a good point in any case. I'll fix that.

> Similarly, instead of having an optional field for "-exe", it is better to have a non-optional, labeled, length-prefixed field.

Well, since we are appending the "-exe" string to something that something that can't contain a `-` or an `x` we should be fine here.

**Comment by michaelwoerister** [comment]
@alexcrichton @FlaPer87 I'm working on putting in the `link-guard` mechanism hinted at at various places and I've run into a problem with the `run-make/relocation-model` test case. 
Some background on the link-guard: At the moment we want to prevent compiling something against one version of a dynamic library and then, at runtime accidentally using a different version of the dynamic library. Before this patch this was impossible because every symbol had the SVH in it and you'd get an error by the dynamic linker when using the wrong version of a dylib. After the patch this isn't the case any more, so we want to add some safety measure to prevent this error case.
The way I have implemented this for locally is the following:
* In every crate that we compile, we emit a function called `__rustc_link_guard_<crate-name>_<crate-svh>`
* The body of this function contains calls to the `__rustc_link_guard` functions of all dependencies.
* An executable contains a call to it's own `__rustc_link_guard` functions.

As a consequence the `__rustc_link_guard` function call graph mirrors the crate graph and the dynamic linker will fail if a wrong dylib is loaded somewhere because its `__rustc_link_guard` function will contain a different SVH in its name.

So far, this works pretty well. However, in the `relocation-model` test I get an error when trying to create a dylib with `-Crelocation-model=static`, which is kind of to be expected because statically compiling a dynamic library does not make much sense (right?). Why does it work without the link-guard? Because the test file being compiled just contains an empty main function and consequently the resulting binary does not contain any relocations. With the link-guard, however, we always reference the `__rustc_link_guard` functions in dependencies and thus linking fails:
```
/usr/bin/ld: /home/mw/rust/x86_64-unknown-linux-gnu/test/run-make/relocation-model/foo.0.o:
relocation R_X86_64_PC32 against symbol `__rustc_link_guard_alloc_0249baca1767b587'
can not be used when making a shared object; recompile with -fPIC
```

My proposed fix for this would be to just remove the cases where a dylib is compiled with `relocation-model=static` since it's not supported in general anyway. But I'd like to get your feedback on this first. Any thoughts?

**Comment by alexcrichton** [comment]
@michaelwoerister yeah that sounds fine to me, it was probably just added via copy/paste and not necessarily pored over much.

**Comment by briansmith** [comment]
> Can you elaborate on that? The value stored as "crate salt" is used as additional input to the hash function we use when generating symbol hashes. I thought that would kind of match up with how the term is used normally.

Usually we (I) would call that "context". Salts are usually random and/or intended to be unguessable by an attacker.

> No, we actually don't. At least as far as I know we just want something that provides good random distribution.

In other words, you don't use an equality comparison of these values as a substitute for a value comparison, then, right? Assuming that is the case, why not just use FNV? If that is not the case, then you are indeed using SHA-256 as a security mechanism.

> Well, since we are appending the "-exe" string to something that something that can't contain a - or an x we should be fine here.

:)

**Comment by michaelwoerister** [comment]
> Usually we (I) would call that "context". Salts are usually random and/or intended to be unguessable by an attacker.

OK, in that case I'll rename it. The term "salt" is only used internally anyway, so it should be OK to use something more unwieldy like "crate_disambiguator".

> In other words, you don't use an equality comparison of these values as a substitute for a value comparison, then, right? Assuming that is the case, why not just use FNV? If that is not the case, then you are indeed using SHA-256 as a security mechanism.

Implicitly we do make an equality comparison: If there is a collision between symbol hashes, you can get linker errors. What we don't care about is whether such collisions can easily be fabricated by an "attacker". What we want is a sufficiently low probability of accidental collisions.

Anyway, more documentation on the respective fields and methods is always a good idea. I'll add those.

**Comment by michaelwoerister** [comment]
@alexcrichton Great, thanks for the quick response!

**Comment by arielb1** [comment]
@michaelwoerister 
@briansmith 

I would really like the SVH to be injective (maybe use RBML?) to avoid annoying matches.

**Comment by michaelwoerister** [comment]
@arielb1 
> I would really like the SVH to be injective (maybe use RBML?) to avoid annoying matches.

That should already be the case in practice, if the hash function used (currently SipHash for the SVH) has a low enough probability of collisions. Can you go into more detail on the motivation of your comment?

**Comment by arielb1** [comment]
@michaelwoerister 

I was talking about the content being hashed over.

There's an annoying issue if, say, you concatenate names together without a separator, so that "foo bar" is the same as "fo obar".

Also, I would prefer to use SHA256 instead of SipHash for collision-resistance - use functions for their correct use!

**Comment by michaelwoerister** [comment]
@arielb1 
Agreed on all points. There probably are some places where something like this could happen in the current implementation.

Note though that this PR is only tangentially related to the SVH, so I'll not try to address these issues here.

**Comment by nikomatsakis** [comment]
@michaelwoerister -- so what are the outstanding blocking issues before this can land? I am thinking that I will rebase https://github.com/rust-lang/rust/pull/32016 on top of it, since the work I am doing there about symbol naming for impl will benefit from the work you've done here (and would conflict otherwise).

**Comment by nikomatsakis** [comment]
Also, I'll steal the review from @Aatch.

**Comment by nikomatsakis** [comment]
(Unless he objects :)

**Comment by nikomatsakis** [comment]
@arielb1 regarding the SVH stuff, I am also using it for https://github.com/rust-lang/rust/pull/32016 to hash the content of HIR nodes and determine what has changed. Obviously in that case I care a lot about conflicts, so I will take a look at fixing the problems you highlighted. Seems straight-forward enough... (There are a few other things I wanted to change anyway)

**Comment by michaelwoerister** [comment]
> so what are the outstanding blocking issues before this can land? 

As of right now, none `:)`

**Comment by nikomatsakis** [comment]
@michaelwoerister very good, will review pronto

**Comment by nikomatsakis** [comment]
@michaelwoerister failure is related to the link guard somehow (from [here](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/2596/steps/compile/logs/stdio)):

```
/c/bot/slave/auto-win-msvc-32-opt/build/mk/target.mk:186: recipe for target 'i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/stamp.arena' failed
Zombie process: "\Device\HarddiskVolume1\dojob.exe"
Zombie process: "\Device\HarddiskVolume1\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\mspdbsrv.exe"
error: linking with `link.exe` failed: exit code: 1120
note: "C:\\Program Files (x86)\\Microsoft Visual Studio 12.0\\VC\\bin\\link.exe" "/LIBPATH:C:\\Program Files (x86)\\Microsoft Visual Studio 12.0\\VC\\lib\\" "/LIBPATH:C:\\Program Files (x86)\\Windows Kits\\8.1\\lib\\winv6.3\\um\\x86" "/NOLOGO" "/NXCOMPAT" "/LARGEADDRESSAWARE" "/SAFESEH" "/LIBPATH:C:\\bot\\slave\\auto-win-msvc-32-opt\\build\\obj\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\\arena-9026086f.0.o" "/OUT:i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\\arena-9026086f.dll" "/DEF:C:\\Users\\RUSTBU~1\\AppData\\Local\\Temp\\rustc.jLSOa3kWrXTl\\lib.def" "i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\\arena-9026086f.metadata.o" "/OPT:REF,ICF" "/DEBUG" "/LIBPATH:i686-pc-windows-msvc/rt" "/LIBPATH:C:/bot/slave/auto-win-msvc-32-opt/build/obj/i686-pc-windows-msvc/llvm/Release/lib" "/LIBPATH:C:\\bot\\slave\\auto-win-msvc-32-opt\\build\\obj\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "/LIBPATH:C:\\bot\\slave\\auto-win-msvc-32-opt\\build\\obj\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "std-9026086f.dll.lib" "ws2_32.lib" "userenv.lib" "shell32.lib" "advapi32.lib" "msvcrt.lib" "/DLL" "/IMPLIB:i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\\arena-9026086f.dll.lib" "compiler-rt.lib"
note:    Creating library i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\arena-9026086f.dll.lib and object i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\arena-9026086f.dll.exp
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_alloc_78bd32441fc9a5a0 referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_libc_5d66cd1beb2e7b25 referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_collections_e8d8d249733f21a9 referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_rustc_unicode_fbebc97273333db7 referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_rand_278eafe13ad9c3cc referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
arena-9026086f.0.o : error LNK2019: unresolved external symbol ___rustc_link_guard_core_0ac0f199e3ecdce4 referenced in function ___rustc_link_guard_arena_e9e0ed99bd4fd87d
i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib\arena-9026086f.dll : fatal error LNK1120: 6 unresolved externals
```

**Comment by michaelwoerister** [comment]
Probably something related to the MSVC toolchain. I'll investigate.

**Comment by eddyb** [comment]
@michaelwoerister Would it be possible, after this change, to avoid storing `item_symbols` in metadata and compute them on-demand?
Last I checked that was the only reason we couldn't write metadata *before* entering `trans` which could allow us to have `.rlib` files usable by other compilations *while* the current compilation is in `trans`, and only delay linking until all the `trans` jobs are done (cc @alexcrichton who knows about my previous attempt at this).

**Comment by michaelwoerister** [comment]
> Would it be possible, after this change, to avoid storing item_symbols in metadata and compute them on-demand?

Yes, that should be possible. Maybe not immediately, as some of the internal symbols still use `gensym_name` but after that is changed (in a subsequent PR) symbol names should be entirely predictable.

> ... have .rlib files usable by other compilations while the current compilation is in trans ...

Oh, neat!

**Comment by michaelwoerister** [comment]
The link-guard symbols where not added to the DLL def file on Windows. Let's give it another try.

**Comment by nikomatsakis** [comment]
failure looks legit http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8420/steps/compile/logs/stdio

**Comment by nikomatsakis** [comment]
@michaelwoerister and I chatted a bit over e-mail. I'm going to close this PR and open one of my own that includes his commits plus a few more.
